sound/soc/sof/mediatek/mt8195/mt8195.c

```diff
@@ -0,0 +1,370 @@
+			ret = -ENOMEM;
+			goto err;
+		}
+	}
+	dev_info(dev, "[ADSP] shared-dram vbase=%p, phy addr :%llx,  size=0x%x\n",
+		 adsp->shared_dram, adsp->pa_shared_dram, shared_size);
+err:
+	return ret;
+}
+
+static int mt8195_dsp_probe(struct snd_sof_dev *sdev)
+{
+	struct platform_device *pdev = container_of(sdev->dev, struct platform_device, dev);
+	struct adsp_priv *priv;
+	struct adsp_mem share_dram;
+	int ret, mailbox_type;
+
+	priv = devm_kzalloc(&pdev->dev, sizeof(*priv), GFP_KERNEL);
+	if (!priv)
+		return -ENOMEM;
+
+	sdev->pdata->hw_pdata = priv;
+	priv->dev = sdev->dev;
+	priv->sdev = sdev;
+
+	priv->adsp = devm_kzalloc(&pdev->dev, sizeof(struct adsp_chip_info),
+				  GFP_KERNEL);
+
+	if (!priv->adsp)
+		return -ENOMEM;
+
+	adsp_info = priv->adsp;
+
+	ret = platform_parse_resource(pdev, adsp_info);
+	if (ret) {
+		dev_err(sdev->dev, "platform_parse_resource failed\n");
+		return ret;
+	}
+
+	ret = platform_parse_clock(&pdev->dev);
+	if (ret) {
+		dev_err(sdev->dev, "platform_parse_clock failed\n");
+		return -EINVAL;
+	}
+
+	pm_runtime_enable(&pdev->dev);
+	pm_runtime_get_sync(&pdev->dev);
+
+	ret = adsp_clock_on(&pdev->dev);
+	if (ret) {
+		dev_err(sdev->dev, "adsp_clock_on fail!\n");
+		goto err_pm_disable;
+	}
+
+	ret = adsp_sram_power_on(sdev->dev, true);
+	if (ret) {
+		dev_err(sdev->dev, "adsp_sram_power_on fail!\n");
+		goto err_pm_disable;
+	}
+
+	ret =  adsp_memory_remap_init(&pdev->dev);
+	if (ret) {
+		dev_err(sdev->dev, "adsp_memory_remap_init fail!\n");
+		goto err_pm_disable;
+	}
+
+	sdev->bar[SOF_FW_BLK_TYPE_IRAM] = devm_ioremap(sdev->dev,
+						       priv->adsp->pa_sram,
+						       priv->adsp->srammsize);
+	if (!sdev->bar[SOF_FW_BLK_TYPE_IRAM]) {
+		dev_err(sdev->dev, "failed to ioremap base %pa size 0x%x\n",
+			&priv->adsp->pa_sram, priv->adsp->srammsize);
+		ret = -EINVAL;
+		goto err_pm_disable;
+	}
+
+#ifdef ADSP_USE_DRAM
+	sdev->bar[SOF_FW_BLK_TYPE_SRAM] = devm_ioremap_wc(sdev->dev,
+							  priv->adsp->pa_dram,
+							  priv->adsp->dramsize);
+	if (!sdev->bar[SOF_FW_BLK_TYPE_SRAM]) {
+		dev_err(sdev->dev, "failed to ioremap base %pa size 0x%x\n",
+			&priv->adsp->pa_dram, priv->adsp->dramsize);
+		ret = -EINVAL;
+		goto err_pm_disable;
+	}
+	adsp_info->va_dram = sdev->bar[SOF_FW_BLK_TYPE_SRAM];
+	mailbox_type = SOF_FW_BLK_TYPE_SRAM;
+#else
+	mailbox_type = SOF_FW_BLK_TYPE_IRAM;
+#endif
+
+	ret =  adsp_shared_base_ioremap(pdev, adsp_info);
+	if (ret) {
+		dev_err(sdev->dev, "adsp_shared_base_ioremap fail!\n");
+		goto err_pm_disable;
+	}
+
+	sdev->mmio_bar = mailbox_type;
+	sdev->mailbox_bar = mailbox_type;
+
+err_pm_disable:
+	pm_runtime_disable(&pdev->dev);
+
+	return ret;
+}
+
+static int mt8195_dsp_remove(struct snd_sof_dev *sdev)
+{
+	struct platform_device *pdev = container_of(sdev->dev, struct platform_device, dev);
+	struct adsp_priv *priv = sdev->pdata->hw_pdata;
+
+	adsp_clock_off(&pdev->dev);
+	adsp_sram_power_on(&pdev->dev, false);
+	pm_runtime_put_sync(&pdev->dev);
```

Comment:
no that's wrong. You should disable pm_runtime here.

Reply:
OK. Currently I still check the pm_runtime flow because I need pm_runtime_get_sync to power on adsp power and clock. I need them to read/write dsp register. Therefore, I need to get_sync to power on adsp power/clock when load firmware, run firmware, and even for IPC.  I am still finding a better method for it. Thanks!

Reply:
@plbossart I need to power on adsp power and clock in the probe function because I need to access dsp registers for some initialization(e.g power on DSP SRAM). Therefore I use get/put sync to ensure not increasing refcount in the probe function. I also do get /put sync paired in these functions : load firmware (get/put sync) , dsp_pre_fw_run(get sync), and machine_register(put sync, I don't put sync in the post_fw_run because I also need adsp power in init_trace, not sure if it is better to get sync in init_trace then put sync in release trace).  With these modification, pm_runtime refcount will be zero after sof_probe function.  But I encounter one problem is that I still need adsp power/clock when load topology but adsp power is power down. Not sure if there is anything I missed or misunderstanding. Could you please give some advice? Thanks a lot!

Reply:
No, that's wrong to auto-resume your own driver. the pm_runtime_get_sync() is used e.g. by ASoC to make sure the DSP is resumed when you start playing audio. It's a utility for *OTHERS* to request the DSP be back in full power mode.

Again, the flow is that on probe you power everything, then mark the DSP as idle and let it suspend if no one else is using the DSP. That determination is done by the pm_runtime framework, not your driver.